app/modules/wifi_eventmon.c

```diff
@@ -79,38 +74,67 @@ static void wifi_event_monitor_handle_event_cb(System_Event_t *evt)
           evt->event == EVENT_SOFTAPMODE_STADISCONNECTED || evt->event == EVENT_SOFTAPMODE_PROBEREQRECVED ||
           evt->event == EVENT_OPMODE_CHANGED)))
   {
-    evt_queue_t *temp = (evt_queue_t*)c_malloc(sizeof(evt_queue_t)); //allocate memory for new queue item
-    temp->evt = (System_Event_t*)c_malloc(sizeof(System_Event_t)); //allocate memory to hold event structure
-    if(!temp || !temp->evt)
-    {
-      luaL_error(lua_getstate(), "wifi.eventmon malloc: out of memory");
-      return;
+    lua_State* L = lua_getstate();
+    if(event_queue_ref == LUA_NOREF){ //if event queue has not been created, create it now
+      lua_newtable(L);
+      event_queue_ref = luaL_ref(L, LUA_REGISTRYINDEX);
     }
-    c_memcpy(temp->evt, evt, sizeof(System_Event_t)); //copy event data to new struct
+    lua_rawgeti(L, LUA_REGISTRYINDEX, event_queue_ref);
 
-    if(wifi_event_queue_head == NULL && wifi_event_queue_tail == NULL)// if queue is empty add item to queue
-    {
-      wifi_event_queue_head = wifi_event_queue_tail = temp;
-      EVENT_DBG("\n\tqueue empty, adding event and posting task\n");
+    System_Event_t* evt_tmp = lua_newuserdata(L, sizeof(System_Event_t));
+    c_memset(evt_tmp, 0, sizeof(System_Event_t));
```

Comment:
this line is really needed as the entire datastriucture is overwritten in the next line

Reply:
>this line is really needed as the entire datastriucture is overwritten in the next line
I assume you meant that it really NOT needed, if so then I can remove it, I added it because I was having a problem due to unreferencing the userdata before processing the event